src/librustc/hir/check_attr.rs

```diff
@@ -93,70 +93,91 @@ struct CheckAttrVisitor<'tcx> {
 impl CheckAttrVisitor<'tcx> {
     /// Checks any attribute.
     fn check_attributes(&self, item: &hir::Item, target: Target) {
-        if target == Target::Fn || target == Target::Const {
-            self.tcx.codegen_fn_attrs(self.tcx.hir().local_def_id(item.hir_id));
-        } else if let Some(a) = item.attrs.iter().find(|a| a.check_name(sym::target_feature)) {
-            self.tcx.sess.struct_span_err(a.span, "attribute should be applied to a function")
-                .span_label(item.span, "not a function")
-                .emit();
-        }
-
         for attr in &item.attrs {
-            if attr.check_name(sym::inline) {
+            let is_valid = if attr.check_name(sym::inline) {
                 self.check_inline(attr, &item.span, target)
             } else if attr.check_name(sym::non_exhaustive) {
                 self.check_non_exhaustive(attr, item, target)
             } else if attr.check_name(sym::marker) {
                 self.check_marker(attr, item, target)
+            } else if attr.check_name(sym::target_feature) {
+                self.check_target_feature(attr, item, target)
+            } else {
+                true
+            };
+
+            if !is_valid {
```

Comment:
Could you move this check outside the loop, and add a test case for reporting errors for multiple invalid attributes (if there's not one already)?

Reply:
I've moved the check, but I haven't added a test because when I did (with two invalid attributes), only one error was reported - I don't think it's my changes doing that.

Reply:
```rust
#![feature(marker_trait_attr)]
#![feature(non_exhaustive)]

#[marker]
#[non_exhaustive]
const FOO: u8 = 0;
```
fails with two errors on [the playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=27edd1b3f033041c69225396b7c15a11). Does it not work if you stop short-circuiting with `is_valid = is_valid && [...]`?

Reply:
Oops, you're right, forgot that it would short-circuit.